feat: Multi config

WillLillis opened the pull request:
This implements multiple configurations as outlined in #153, supporting both a default configuration as well as any number of project configurations for sub-directories. For example, given a project with the following structure:

```shell
.
├── .asm-lsp.toml
├── compile_flags.txt
└── sub
    ├── arm.s
    ├── gas.s
    └── subsub
        ├── arm.s
        └── gas.s
```

and the following `.asm-lsp.toml` file:

```toml
[[projects]]
path = "sub"

[projects.assemblers]
gas = true

[projects.instruction_sets]
arm = true

[projects.opts]
compiler = "zig" # need "cc" as the first argument in `compile_flags.txt`

[[projects]]
path = "sub/subsub"

[projects.assemblers]
gas = true

[projects.instruction_sets]
x86 = true
x86_64 = true

[projects.opts]
compiler = "zig" # need "cc" as the first argument in `compile_flags.txt`
```

only ARM-related features will be available in the `sub` directory, but only x86/x86-64 features will be available in the `subsub` directory. This also works as one would expect with sibling directories. The `path` field of `[[projects]]` can be relative to the project root or an absolute path.

This change required some larger changes than I originally expected, since we can't just filter out the documentation we don't want at load time anymore. In my initial implementation for this, I had to throw away a few generic functions, which blew up the code size a bit. I'll take another pass on cleaning these portions up.

Any and all feedback is very appreciated. A few TODOs/ open questions to start:

- [x] Fix CI
- [x] Make the `opts` config field optional
- [x] Handle multiple project configurations pointing to the same directory or file
- [x] General code cleanup as mentioned above
- [x] Improved documentation of this feature in README.md
- [x] Add optional `compile_cmd` to `project` configs
- [x] Restrict `instruction_set` and `assembler` config fields to only hold compatible values
- [x] Other suggestions/ ideas as they come up
- Behavior of the server when there isn't a default configuration provided and no project configuration applies to a source file. ~~Currently, the server defaults to an "empty" config with all ISA/Assembler-specific features (including diagnostics) turned off. This still allows for some functionality (e.g. name demangling, symbol goto definition, etc.).~~ Enforce some reasonable defaults.
- ~~Should `compile_flags.txt`/`compile_commands.json` also be treated in the same way as the config file?~~ Provide `compile_flags_txt` field for projects.

This will stay as a draft until #156 is merged. 

cc @Ultra-Code @lu-zero

Closes #153

lu-zero commented:
Thank you a lot for working on it!

lu-zero commented:
> Should compile_flags.txt/compile_commands.json also be treated in the same way as the config file?
Ideally they are generated in the build directory anyway, so potentially adding an env var so the IDE can tell where to look at it on a path different from `{project root}/build` would be already great.

Ultra-Code commented:
I will test out the changes in this PR and give better suggestions later but the first change I think we should make is to change `[[projects]] to [[project]]` as `[[project]]` is an array of tables  and using the plural `projects` could be deceptive and cause users to think they can add multiple `project` in `projects`

WillLillis commented:
> I'm looking at the `InstructionSets` struct and with this PR in progress, I think it makes sense to restrict the `instruction set` to only compatible architectures. .ie for a project on `x86` should be able to enable `x86_64` also but not `riscv` or `arm64` because those can be in other `[[project]]` sections, what do you think?

That seems pretty reasonable. One use case I see that goes against this is someone cross compiling a project from a higher level language and inspecting the resulting assembly files (say one is x86, the other ARM) all into the same `build` directory. To cover this situation, we could also allow the `path` field for each `project` entry to specify individual files instead of just directories. Thoughts?

Also, do you think it would make sense to similarly restrict the `assemblers` field to a single value?

Ultra-Code commented:
Nice idea, being able to specify a file path instead of a directory would fix that use case. And yes I am on board with the idea of specifying a single assembler because, to the best of my knowledge, assemblers aren't compatible .ie assembly written for gas, or go won't work with nasm or masm.

WillLillis commented:
> > Should compile_flags.txt/compile_commands.json also be treated in the same way as the config file?
> 
> Ideally they are generated in the build directory anyway, so potentially adding an env var so the IDE can tell where to look at it on a path different from `{project root}/build` would be already great.

The project currently looks in both the project root and `{project root}/build` for these files. Looking at this again, `compile_commands.json` already handles multiple projects well, so there's no real need for a per-project instance of that file. I think it does make sense to allow for per-project `compile_flags.txt` files.

I'd be open to adding support for an environment variable to provide additional search paths for these files. A few ideas on it's behavior:
- If the path given by the environment variable is absolute, it would just check in that directory.
- If it's relative, we could:
  - append it to the project root or root build folder to look for `compile_commands.json` or `compile_flags.txt`
  - append it to each `project` path to search for `compile_flags.txt`

Ultra-Code commented:
By `build` directory is that supposed to be the `target` dir for `Rust` and the `zig-out` for `Zig` ?

Yes, Per project `compile_flags.txt` would be needed as each would be building for different architectures.

WillLillis commented:
> By `build` directory is that supposed to be the `target` dir for `Rust` and the `zig-out` for `Zig` ?

Currently `build` directory just refers to a literal folder named `build` in the project's root directory. I adopted this convention from the llvm project's [docs](https://clang.llvm.org/docs/JSONCompilationDatabase.html#build-system-integration) but am open to amending it. 

> Yes, Per project `compile_flags.txt` would be needed as each would be compiling for different architectures or we can add the ability to specify the full compiler invocation in `.asm-lsp.toml` per project, removing the need for a `compile_flags.txt`.

I think it still makes sense to support `compile_flags.txt`, since it's an established standard that many other tools can produce, use, etc. Allowing the user to specify the full compiler invocation in `.asm-lsp.toml` on a per project basis seems like a nice change though. I can look into adding that in this PR.

> Can you throw more light on the environment variable use case, as with the `path` field been able to take either a `directory` or `file` with the addition of a per project `opt` table and potentially the ability to specify the full compiler invocation in `opt.compiler` of a project, the additional environment variable doesn't seem very useful.

I'll defer to @lu-zero to clarify their use-case here.

> And correct me if am wrong, but from my understanding, any `IDE` or `Editor` would communicate with the `lsp` to get diagnostics and doesn't necessarily need to know about these files as the `lsp` takes care of it.

Correct. The IDE/editor's LSP client communicates the project's root directory to `asm-lsp` at start time, and after that any file detection is up to us.

WillLillis commented:
Actually taking another look at this, trying to allow for `compile_flags.txt` in sub-directories is fairly messy when they have to interact with the other ways compiler invocations can be specified. I think it makes more sense to allow for a root `compile_flags.txt` as is already the case, and using @Ultra-Code's idea in using the project configs to specify per-project compiler invocations as an optional way to override any specified at the root level.

lu-zero commented:
> I'll defer to @lu-zero to clarify their use-case here.

The editor needs to cooperate with the build system and the lsp, I'm afraid there isn't a standardized way to pass around information among the 3, I tend to use `.build` and `.build-{target}` while developing on C projects. So my use-case would be to override the clangd-derived `build` while I'm using vim/neovim.

Ultra-Code commented:
@lu-zero I am not sure I still fully understand your use case, as looking at the [comp db format](https://clang.llvm.org/docs/JSONCompilationDatabase.html#format) I'm sure depending on how you invoke the command to create the `comp db file` .eg `cmake -S . -B .build -DCMAKE_EXPORT_COMPILE_COMMANDS=ON`, your build directory would be appropriately set to either the `.build` or `.build-{target}` in the directory field of the command object as stated in the documentation.

Then you can symlink the generated `compile_command.json` from `.build` into your `project root` to make it easy for tools to find it.

And if maybe you are using a plugin that tries to emulate an lsp like [none-ls](https://github.com/nvimtools/none-ls.nvim/blob/main/doc/BUILTINS.md#clazy) you still would need the `comp db` with the right build directory set so I am still a little bit confused how that extra environment variable would help the editor if it isn't going through `compile_command.json` or `compile_flags.txt`

WillLillis commented:
Alrighty I believe I've addressed most comments. I can address adding support for a build directory environment variable once that discussion is resolved. I'm planning on fixing up the temporary lint issues and merging in #156 tomorrow after work. After that, rebasing this should be interesting :joy:.

Being able to iterate on this PR with steady feedback has been incredibly helpful. Please keep the requests/suggestions coming. :)

Ultra-Code commented:
I am reviewing your code in this branch and I think there are some changes we can make moving forward to help with ease of reasoning about some of the functions and maintainability.

With the change, I suggested in mind
> I'm looking at the InstructionSets struct and with this PR in progress, I think it makes sense to restrict the instruction set to only compatible architectures. .ie for a project on x86 should be able to enable x86_64 also but not riscv or arm64 because those can be in other [[project]] sections, what do you think?

I think it makes seen to switch to an enum for the `InstructionSet` so that we can switch on the `Architectures` and only execute code relevant to the architecture set in the config and provide useful error messages when incompatible `Arch`s are enabled. This would be a breaking change to user configs but I think it would come with some perf improvements and code simplification as we won't be doing work as though all architectures have been enabled and can all be usable which isn't true in reality. And doing the same for the assemblers.


I kinda agree with Clippy on the too many argument Lint
Only a few Arch support 8 arguments passed in registers .ie Aarch64, Riscv64, PowerPC, and MISP. The venerable x86_64 supports 6 arguments, x86 supports 2 arguments, and arm32 supports 4 arguments. I can see this reaching 12 soon 😄 . In all, I think it is a good idea to start reasoning about the parameters that are related and putting them in a struct.

I see in a todo that you plan to do some compatibility work for `arm32/arm63` but most new aarch64 processors aren't [compatible with arm32](https://newsroom.arm.com/blog/64-bit) so I won't advise that.

I will try to get a patch set in for some of these changes.

Ultra-Code commented:
🤣 I am by no means an ASM expert 😆 but I am currently on an ASM learning journey.

Ultra-Code commented:
Got a patch cooking in [restructure](https://github.com/Ultra-Code/asm-lsp/tree/restructure) with some of the ideas stated above, still WIP though.

Ultra-Code commented:
@WillLillis I am done with the ideas I had in mind for the multi-config support, I will leave the parameters restructuring to you.

We can now have a beautiful multi-config setup like this and still be efficient :wink:

```toml
version = "0.1"
[default_config]
assembler = { gas = true }
instruction_set = { arm = true }

[opts]
compiler = "zig"
diagnostics = true
default_diagnostics = true

[[project]]
path = "aarch64"
instruction_set = { arm = true }
assembler = { gas = true }
[project.opts]
compiler = "zig"
compile_flags_txt = [
  "cc",
  "-x",
  "assembler-with-cpp",
  "-g",
  "-Wall",
  "-Wextra",
  "-pedantic",
  "-pedantic-errors",
  "-std=c2y",
  "-target",
  "aarch64-linux-musl",
]

[[project]]
path = "x86_64"
assembler = { gas = true }
instruction_set = { x86_64 = true }
```

I came across the `empty fn()` which seem to be the 9th task issue you where asking for suggestions on, I don't think we should allow such cases. In my opinion it over complicates things and make it difficult for me to tell the behavior of the server just by looking at the `.asm-lsp.toml` config. So I suggest we make a user specify the `assembler` and `instruction_set` for each project or set the `default_config` options so that options that aren't set in the project could use those from `default_config`

Pertaining to the last point, I changed `compile_cmd` to `compile_flags_txt` to better model what exactly that config option is doing and how an actual `compile_flags.txt` would be. You may give feedback on my changes.

WillLillis commented:
@Ultra-Code 
> @WillLillis I am done with the ideas I had in mind for the multi-config support, I will leave the parameters restructuring to you.
> 
> 
> 
> We can now have a beautiful multi-config setup like this and still be efficient :wink:
> 
> 
> 
> ```toml
> 
> version = "0.1"
> 
> [default_config]
> 
> assembler = { gas = true }
> 
> instruction_set = { arm = true }
> 
> 
> 
> [opts]
> 
> compiler = "zig"
> 
> diagnostics = true
> 
> default_diagnostics = true
> 
> 
> 
> [[project]]
> 
> path = "aarch64"
> 
> instruction_set = { arm = true }
> 
> assembler = { gas = true }
> 
> [project.opts]
> 
> compiler = "zig"
> 
> compile_flags_txt = [
> 
>   "cc",
> 
>   "-x",
> 
>   "assembler-with-cpp",
> 
>   "-g",
> 
>   "-Wall",
> 
>   "-Wextra",
> 
>   "-pedantic",
> 
>   "-pedantic-errors",
> 
>   "-std=c2y",
> 
>   "-target",
> 
>   "aarch64-linux-musl",
> 
> ]
> 
> 
> 
> [[project]]
> 
> path = "x86_64"
> 
> assembler = { gas = true }
> 
> instruction_set = { x86_64 = true }
> 
> ```

Sounds great! One question regarding this formulation of the config. For selecting the assemblers and instruction sets, what advantages do you see for the above style with over allowing a more direct assignment, i.e.:

```toml
assembler = "Gas"
instruction_set = "x86"
```

One obvious plus I see to your implementation is it's clearer to users regarding how to enable multiple "compatible" arches (i.e. x86 and x86-64). One disadvantage I see is it would also be easier to assign multiple incompatible arches. I suppose this worry could also just be addressed in adding a note in the README. 
I could just be misunderstanding, though. Maybe it doesn't make sense to have both x86 and x86-64 enabled at the same time?  

> I came across the `empty fn()` which seem to be the 9th task issue you where asking for suggestions on, I don't think we should allow such cases. In my opinion it over complicates things and make it difficult for me to tell the behavior of the server just by looking at the `.asm-lsp.toml` config. So I suggest we make a user specify the `assembler` and `instruction_set` for each project or set the `default_config` options so that options that aren't set in the project could use those from `default_config`

I'm fine with this. Requiring a default config simplifies things and removes some unintuitive behavior. 

We do still need to cover behavior if there isn't a config file present, though. One idea I had was to set the default architecture by detecting the host system's architecture. If it can't be detected, we can also use some fallback value. I think Gas should work fine as a fallback assembler.

> Pertaining to the last point, I changed `compile_cmd` to `compile_flags_txt` to better model what exactly that config option is doing and how an actual `compile_flags.txt` would be. You may give feedback on my changes.

This sounds good to me. I think making the behavior between normal compile_flags.txt files and our config schema is a positive.

Thanks so much for all the feedback and spending time working on this! I haven't had as much free time as I'd like the last few days to get this PR's changes on top of the latest master, but I am working on it. :) I'll probably have some more questions as I get further into integrating your changes and mine on top of the master branch, but those are the thoughts I have for now.

WillLillis commented:
Had to squash all of my previous commits to actually get through the rebase, but everything is back on top of master now. I'll start cleaning up my work and adding your patches in next :)

Ultra-Code commented:
>Sounds great! One question regarding this formulation of the config. For selecting the assemblers and instruction sets, what advantages do you see for the above style with over allowing a more direct assignment

The first thing that made me gravitate towards the solution above is a config error I got when I used `assember = "go"` but arguable this could have been changed easily by just serializing in lower case but I also thought the previous syntax for the config was simple and neat,and didn't think we should change that structure just because some underlying details have changes (.ie change from struct to enum)

>One obvious plus I see to your implementation is it's clearer to users regarding how to enable multiple "compatible" arches (i.e. x86 and x86-64). One disadvantage I see is it would also be easier to assign multiple incompatible arches. I suppose this worry could also just be addressed in adding a note in the README.
I could just be misunderstanding, though. Maybe it doesn't make sense to have both x86 and x86-64 enabled at the same time?

It actually doesn't support multiple `archs` or `assemblers` as all the supported combinations are covered in the backing enum. An informative error is given to the user when they mistakenly do that in their config.

I haven't really looked deep into what the x86-64 `mnemonics` and `registers` contain but `x86-64` is a super set of `x86` so if the implementation duplicated the common parts between the two then then `x86_and_x86-64` arch wouldn't be needed

>I'm fine with this. Requiring a default config simplifies things and removes some unintuitive behavior.
We do still need to cover behavior if there isn't a config file present, though. One idea I had was to set the default architecture by detecting the host system's architecture. If it can't be detected, we can also use some fallback value. I think Gas should work fine as a fallback assembler.

I agree with you, that would be intuitive.

WillLillis commented:
> > Sounds great! One question regarding this formulation of the config. For selecting the assemblers and instruction sets, what advantages do you see for the above style with over allowing a more direct assignment
> 
> The first thing that made me gravitate towards the solution above is a config error I got when I used `assember = "go"` but arguable this could have been changed easily by just serializing in lower case but I also thought the previous syntax for the config was simple and neat,and didn't think we should change that structure just because some underlying details have changes (.ie change from struct to enum)

That's fair, I definitely don't think the config values should be case sensitive. However, having to explicitly set an architecture to `true` seems a little indirect to me. I think it's clearer/simpler to have a direct assignment with the enum implementation. That way, a number of invalid config states just aren't able to be represented (i.e. multiple incompatible architectures enabled at the same time), rather than checking for them after deserialization. My experience is that most users don't look in the log file, so issuing a warning about a bad config will most likely be ignored (unfortunately :(). Since we're making a breaking change here anyway, I think it's a great opportunity to restructure things a bit. I will definitely spend some more time thinking this over though.

> > One obvious plus I see to your implementation is it's clearer to users regarding how to enable multiple "compatible" arches (i.e. x86 and x86-64). One disadvantage I see is it would also be easier to assign multiple incompatible arches. I suppose this worry could also just be addressed in adding a note in the README.
> > I could just be misunderstanding, though. Maybe it doesn't make sense to have both x86 and x86-64 enabled at the same time?
> 
> It actually doesn't support multiple `archs` or `assemblers` as all the supported combinations are covered in the backing enum. An informative error is given to the user when they mistakenly do that in their config.
> 
> I haven't really looked deep into what the x86-64 `mnemonics` and `registers` contain but `x86-64` is a super set of `x86` so if the implementation duplicated the common parts between the two then then `x86_and_x86-64` arch wouldn't be needed
> 
> > I'm fine with this. Requiring a default config simplifies things and removes some unintuitive behavior.
> > We do still need to cover behavior if there isn't a config file present, though. One idea I had was to set the default architecture by detecting the host system's architecture. If it can't be detected, we can also use some fallback value. I think Gas should work fine as a fallback assembler.
> 
> I agree with you, that would be intuitive.

Sounds great! Really appreciate your thoughts on this. I'll be picking away on working in your improvements over the next few days.

WillLillis commented:
One other thing I'm considering is having the server crash when it reads in an invalid configuration file. The current behavior is to fall back to a default config, but this can be unintuitive since the failure is silent (unless you check the logs).

Ultra-Code commented:
>One other thing I'm considering is having the server crash when it reads in an invalid configuration file. The current behavior is to fall back to a default config, but this can be unintuitive since the failure is silent (unless you check the logs).

I think this should be the way to go, as using the default config would lead to unexpected behavior on the user side. But I don't know if we can crash the server in a way that the user would see in there editor info section instead of just with logs .eg `asm-lsp crashed due to invalid config` instead of `lsp server exited`, to prevent users from filing issues when its just incorrect config.
Also we can add a big sign in the `README.md` that the server would crash when given incorrect config.

lu-zero commented:
Please never crash, close and return an error value :)

WillLillis commented:
> >One other thing I'm considering is having the server crash when it reads in an invalid configuration file. The current behavior is to fall back to a default config, but this can be unintuitive since the failure is silent (unless you check the logs).
> 
> 
> 
> I think this should be the way to go, as using the default config would lead to unexpected behavior on the user side. But I don't know if we can crash the server in a way that the user would see in there editor info section instead of just with logs .eg `asm-lsp crashed due to invalid config` instead of `lsp server exited`, to prevent users from filing issues when its just incorrect config.

I'd really like to be able to display a message to the user regarding why the server is exiting, but I'm not aware of any parts of the LSP protocol we could use to send this message. I'll spend some time looking into this. 

> Also we can add a big sign in the `README.md` that the server would crash when given incorrect config.

Definitely

Ultra-Code commented:
@WillLillis I have rebase my branch [restructure](https://github.com/Ultra-Code/asm-lsp/tree/restructure) onto yours to make integration easy.

WillLillis commented:
> Please never crash, close and return an error value :)

Sorry, poor wording on my part. Locally at least, panics return an exit code of 101 to the OS. For this we could explicitly return some other non-zero code. Maybe settle on -1 for invalid configs, and we can use other exit codes for other specific issues as they come up.

WillLillis commented:
> @WillLillis I have rebase my branch [restructure](https://github.com/Ultra-Code/asm-lsp/tree/restructure) onto yours to make integration easy.

Thank you! I will be spending some time on this tonight :)

WillLillis commented:
Invalid configurations now have significantly improved user visibility. Here's an example:

![InvalidConfig](https://github.com/user-attachments/assets/020f28e2-e636-4ca3-9b2e-696f2841b3fc)

Starting on working in some of Ultra's additions now :)

WillLillis commented:
I won't be able to work on this much this weekend, but I'll start up again on Monday. I still need to do some more general cleanup, lots of testing around diagnostics-related features (there's a few different configurations that need checking, and we don't have integration tests to cover this), updating the config section of the readme some more, and I still need to cut down on the number of arguments passed to many functions.

WillLillis commented:
Got back in a little earlier than expected! I think this is ready to be merged in. @Ultra-Code If you have time to look this over within the next week or so, another set of eyes would be greatly appreciated. :slightly_smiling_face:

Ultra-Code commented:
At first glance, this PR looks good but there are some improvement that could be made
1. using more match statement at appropriate place
2. Reducing code duplication by refactoring and reusing functionality

Give me a day or two to test this out locally and revert with any further changes/suggestions

Ultra-Code commented:
Tested this PR out, found two new bug which I have stated above
1. arm64 isn't serialized as such
2. result of `match` in `get_compile_cmd_for_path` isn't returned resulting in `None` used all the time

WillLillis commented:
> Tested this PR out, found two new bug which I have stated above
> 
> 1. arm64 isn't serialized as such
> 
> 2. result of `match` in `get_compile_cmd_for_path` isn't returned resulting in `None` used all the time

Thanks for taking a look! I'm not sure what would be causing the first issue, but it's probably a quick fix. The second issue is due to some mishandled interaction between `get_compile_cmd_for_path` and `get_default_compile_cmd`. This logic has gotten a little spaghetti-y, so I'll spend some time refactoring it tonight.

WillLillis commented:
Threw away `get_compile_cmd_for_path` and replaced it with `get_compile_cmd_for_req`. The former was returning `None` every time, but after looking at it again, it wasn't applying flags from a `compile_flags.txt` file with a compiler specified in `.asm-lsp.toml`. I'll take a fresh look at this tomorrow (as well as your previous two comments), but I think the behavior is correct now. 

Given how tricky these issues can be to debug, I think it would be a good idea to set up some additional testing at some point. I'll keep this in mind as a future TODO.

WillLillis commented:
Spent some time cleaning things up. I feel pretty ok with the changes to the server's core, and am very happy with how short `setup_registers`, `setup_instructions`, and `setup_directives` are now. Since this is such a big change, I'm going to let this sit for a few days before merging to give people (and myself after I've gotten some sleep :laughing:) a chance to check things over.
